core/src/storage/lazy/lazy_chunk.rs

```diff
@@ -0,0 +1,384 @@
+    ///
+    /// # Note
+    ///
+    /// Only loads a value if `key` is set and if the value has not been loaded yet.
+    /// Returns the freshly loaded or already loaded entry of the value.
+    ///
+    /// # Safety
+    ///
+    /// This function has a `&self` receiver while returning an `Option<*mut T>`
+    /// which is unsafe in isolation. The caller has to determine how to forward
+    /// the returned `*mut T`.
+    ///
+    /// # Panics
+    ///
+    /// - If the lazy chunk is in an invalid state that forbids interaction.
+    /// - If the lazy chunk is not in a state that allows lazy loading.
+    fn lazily_load(&self, index: Index) -> *mut Entry<T> {
+        let key = match self.key {
+            Some(key) => key,
+            None => panic!("cannot load lazily in this state"),
+        };
+        // SAFETY: We have put the whole `cached_entries` mapping into an
+        //         `UnsafeCell` because of this caching functionality. The
+        //         trick here is that due to using `Box<T>` internally
+        //         we are able to return references to the cached entries
+        //         while maintaining the invariant that mutating the caching
+        //         `BTreeMap` will never invalidate those references.
+        //         By returning a raw pointer we enforce an `unsafe` block at
+        //         the caller site to underline that guarantees are given by the
+        //         caller.
+        let cached_entries = unsafe { &mut *self.cached_entries.get() };
+        use ink_prelude::collections::btree_map::Entry as BTreeMapEntry;
+        match cached_entries.entry(index) {
+            BTreeMapEntry::Occupied(occupied) => &mut **occupied.into_mut(),
+            BTreeMapEntry::Vacant(vacant) => {
+                let value = match crate::env::get_contract_storage::<T>(key + index) {
+                    Some(new_value) => {
+                        Some(new_value.expect("could not decode lazily loaded value"))
+                    }
+                    None => None,
+                };
+                &mut **vacant.insert(Box::new(Entry {
+                    value,
+                    mutated: core::cell::Cell::new(false),
+                }))
+            }
+        }
+    }
+
+    /// Lazily loads the value at the given index.
+    ///
+    /// # Note
+    ///
+    /// Only loads a value if `key` is set and if the value has not been loaded yet.
+    /// Returns the freshly loaded or already loaded entry of the value.
+    ///
+    /// # Panics
+    ///
+    /// - If the lazy chunk is in an invalid state that forbids interaction.
+    /// - If the lazy chunk is not in a state that allows lazy loading.
+    fn lazily_load_mut(&mut self, index: Index) -> &mut Entry<T> {
+        // SAFETY: Dereferencing the raw-pointer here is safe since we
+        //         encapsulated this whole call with a `&mut self` receiver.
+        //         Also `Entry` instances are all pinned, so mutating the
```

Comment:
Pinned how? The type is now `BTreeMap<Index, Box<Entry<T>>>`

Reply:
They _were_ pinned but now only the `Box` and the handling around it makes sure that elements are not moved.